File: Lab 4/Java/Assignment7.java
import java.util.*;

public class Assignment7 {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        if (!sc.hasNextInt())
            return;
        // Assignment7.py output Step 352 Block 3:
        // "rows_var, columns_var, knights_var = map(int, input().split())"
        // NO t_var loop.

        int r = sc.nextInt();
        int c = sc.nextInt();
        int k = sc.nextInt();

        Set<String> positions = new HashSet<>();
        int[][] knightPos = new int[k][2];

        for (int i = 0; i < k; i++) {
            int rr = sc.nextInt();
            int cc = sc.nextInt();
            knightPos[i][0] = rr;
            knightPos[i][1] = cc;
            positions.add(rr + "," + cc);
        }

        boolean attack = false;
        for (int i = 0; i < k; i++) {
            int cr = knightPos[i][0];
            int cc = knightPos[i][1];

            int[][] moves = {
                    { cr - 2, cc - 1 }, { cr - 2, cc + 1 },
                    { cr - 1, cc - 2 }, { cr - 1, cc + 2 },
                    { cr + 1, cc - 2 }, { cr + 1, cc + 2 },
                    { cr + 2, cc - 1 }, { cr + 2, cc + 1 }
            };

            for (int[] move : moves) {
                if (positions.contains(move[0] + "," + move[1])) {
                    attack = true;
                    break;
                }
            }
            if (attack)
                break;
        }

        if (attack)
            System.out.println("YES");
        else
            System.out.println("NO");
    }
}

File: Lab 4/Assignment7.py
import sys  # fast stdin

input = sys.stdin.readline  # quick line reads

rows_var, columns_var, knights_var = map(int, input().split())  # board size and knight count

positions_var = {}  # map of occupied cells -> index (presence set)


i_var = 0  # knight counter

while i_var < knights_var:  # read knight positions

    pos_var = tuple(map(int, input().split()))  # (row, col)

    positions_var[pos_var] = i_var  # store occupancy

    i_var = i_var + 1

flag_var = True  # assume no attacks initially

for pos_var in positions_var.keys():  # check each knight

    moves_var = [  # all 8 knight moves from current square
        (pos_var[0] - 2, pos_var[1] - 1),
        (pos_var[0] - 2, pos_var[1] + 1),
        (pos_var[0] - 1, pos_var[1] - 2),
        (pos_var[0] - 1, pos_var[1] + 2),
        (pos_var[0] + 1, pos_var[1] - 2),
        (pos_var[0] + 1, pos_var[1] + 2),
        (pos_var[0] + 2, pos_var[1] - 1),
        (pos_var[0] + 2, pos_var[1] + 1)
    ]

    j_var = 0

    while j_var < len(moves_var):  # test each reachable square

        if moves_var[j_var] in positions_var:  # another knight found

            flag_var = False  # attack exists

            break

        j_var = j_var + 1
    if not flag_var:  # early exit if attack detected

        break

if flag_var:  # no attacks

    print("NO")
else:
    
    print("YES")